File: backend/app/routers/financeiro.py
"""Gestão financeira de alunos — cobrancas e faturamento."""
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query

from app.dependencies import get_current_personal_id
from app.models.financeiro import CobrancaConfigIn, NovaCobrancaIn, RegistrarPagamentoIn
from app.services import authz, financeiro_service

router = APIRouter(prefix="/v1/alunos/{aluno_id}/financeiro", tags=["financeiro"])

# Painel financeiro do personal (visão de carteira, agregada) — sem aluno_id no path.
painel_router = APIRouter(prefix="/v1/financeiro", tags=["financeiro-painel"])


def _aluno_do_personal(aluno_id: str, personal_id: str = Depends(get_current_personal_id)):
    authz.authorize_aluno(personal_id, aluno_id)
    return personal_id


@painel_router.get("/resumo")
def resumo_financeiro(
    mes: Optional[str] = Query(default=None, pattern=r"^\d{4}-\d{2}$"),
    personal_id: str = Depends(get_current_personal_id),
):
    return financeiro_service.resumo(personal_id, mes)


@painel_router.get("/recebiveis")
def listar_recebiveis(
    status: Optional[str] = Query(default=None, pattern="^(PENDENTE|VENCIDA)$"),
    personal_id: str = Depends(get_current_personal_id),
):
    return {"items": financeiro_service.listar_recebiveis(personal_id, status)}


@painel_router.get("/pagamentos-recentes")
def listar_pagamentos_recentes(
    limit: int = Query(default=20, ge=1, le=100),
    personal_id: str = Depends(get_current_personal_id),
):
    return {"items": financeiro_service.listar_pagamentos_recentes(personal_id, limit)}


@router.get("/config")
def get_config(aluno_id: str, personal_id: str = Depends(_aluno_do_personal)):
    cfg = financeiro_service.get_config(personal_id, aluno_id)
    return cfg or {}


@router.put("/config")
def set_config(aluno_id: str, body: CobrancaConfigIn,
               personal_id: str = Depends(_aluno_do_personal)):
    return financeiro_service.set_config(personal_id, aluno_id, body.model_dump())


@router.get("/cobrancas")
def listar_cobrancas(
    aluno_id: str,
    limit: int = Query(default=50, ge=1, le=100),
    cursor: Optional[str] = None,
    status: Optional[str] = Query(default=None, pattern="^(PENDENTE|PAGA|VENCIDA)$"),
    personal_id: str = Depends(_aluno_do_personal),
):
    items, next_cursor = financeiro_service.listar_cobrancas(
        personal_id, aluno_id, limit=limit, cursor=cursor, status=status)
    return {"items": items, "next_cursor": next_cursor}


@router.post("/cobrancas", status_code=201)
def criar_cobranca(aluno_id: str, body: NovaCobrancaIn,
                   personal_id: str = Depends(_aluno_do_personal)):
    return financeiro_service.criar_cobranca_manual(personal_id, aluno_id, body.model_dump())


@router.patch("/cobrancas/{cobranca_id}/pagar")
def registrar_pagamento(aluno_id: str, cobranca_id: str,
                         body: RegistrarPagamentoIn,
                         personal_id: str = Depends(_aluno_do_personal)):
    result = financeiro_service.registrar_pagamento(
        personal_id, aluno_id, cobranca_id, body.model_dump())
    if result is None:
        raise HTTPException(404, "Cobrança não encontrada")
    return result


@router.delete("/cobrancas/{cobranca_id}", status_code=204)
def cancelar_cobranca(aluno_id: str, cobranca_id: str,
                       personal_id: str = Depends(_aluno_do_personal)):
    ok = financeiro_service.cancelar_cobranca(personal_id, aluno_id, cobranca_id)
    if not ok:
        raise HTTPException(404, "Cobrança não encontrada ou já paga")

File: frontend/src/api/financeiro.ts
import { api } from './client'
import { alunoClient } from './alunoClient'
import type { Cobranca, CobrancaConfig, CobrancaConfigIn, NovaCobrancaIn, RegistrarPagamentoIn, CobrancaStatus } from '../types'

export interface CobrancasResponse {
  items: Cobranca[]
  next_cursor?: string
}

export interface MercadoPagoStatus {
  configurado: boolean
  configurado_em?: string
}

// ── Painel financeiro (visão de carteira do personal) ──────────────────────────
export interface FinanceiroResumo {
  mes: string
  recebido_valor: number
  pago_count: number
  a_receber_valor: number
  pendente_count: number
  vencido_valor: number
  vencido_count: number
  mrr: number
  mp_configurado: boolean
  mp_liquido: number | null
  mp_taxa: number | null
  historico: { mes: string; recebido: number }[]
}

/** Cobrança em aberto/paga enriquecida com o nome do aluno (endpoints agregados). */
export type CobrancaComAluno = Cobranca & { aluno_nome: string }

export const financeiroPainelApi = {
  getResumo: (mes?: string) =>
    api.get<FinanceiroResumo>('/v1/financeiro/resumo', { params: mes ? { mes } : undefined }).then((r) => r.data),

  listRecebiveis: (status?: 'PENDENTE' | 'VENCIDA') =>
    api.get<{ items: CobrancaComAluno[] }>('/v1/financeiro/recebiveis', { params: status ? { status } : undefined })
      .then((r) => r.data.items),

  listPagamentosRecentes: (limit = 20) =>
    api.get<{ items: CobrancaComAluno[] }>('/v1/financeiro/pagamentos-recentes', { params: { limit } })
      .then((r) => r.data.items),
}

export const financeiroApi = {
  getConfig: (alunoId: string) =>
    api.get<CobrancaConfig>(`/v1/alunos/${alunoId}/financeiro/config`).then((r) => r.data),

  setConfig: (alunoId: string, body: CobrancaConfigIn) =>
    api.put<CobrancaConfig>(`/v1/alunos/${alunoId}/financeiro/config`, body).then((r) => r.data),

  listCobrancas: (alunoId: string, params?: { status?: CobrancaStatus; cursor?: string }) =>
    api.get<CobrancasResponse>(`/v1/alunos/${alunoId}/financeiro/cobrancas`, { params }).then((r) => r.data),

  createCobranca: (alunoId: string, body: NovaCobrancaIn) =>
    api.post<Cobranca>(`/v1/alunos/${alunoId}/financeiro/cobrancas`, body).then((r) => r.data),

  registrarPagamento: (alunoId: string, cobrancaId: string, body: RegistrarPagamentoIn) =>
    api.patch<Cobranca>(`/v1/alunos/${alunoId}/financeiro/cobrancas/${cobrancaId}/pagar`, body).then((r) => r.data),

  cancelarCobranca: (alunoId: string, cobrancaId: string) =>
    api.delete(`/v1/alunos/${alunoId}/financeiro/cobrancas/${cobrancaId}`),

  getMpConfig: () =>
    api.get<MercadoPagoStatus>('/v1/config/mercadopago').then((r) => r.data),

  setMpConfig: (accessToken: string) =>
    api.put('/v1/config/mercadopago', { access_token: accessToken }),

  deleteMpConfig: () =>
    api.delete('/v1/config/mercadopago'),
}

export interface PixResponse {
  payment_id: string
  qr_code: string
  qr_code_base64: string
  expires_at: string
}

export interface PixStatusResponse {
  payment_id: string
  status: 'pending' | 'approved' | 'rejected' | 'cancelled' | string
  valor_liquido: number
  taxa: number
}

export const alunoFinanceiroApi = {
  listCobrancas: () =>
    alunoClient.get<CobrancasResponse>('/v1/aluno/financeiro').then((r) => r.data),

  getMpConfigurado: () =>
    alunoClient.get<{ configurado: boolean }>('/v1/aluno/financeiro/mp/configurado').then((r) => r.data),

  criarPix: (cobrancaId: string) =>
    alunoClient.post<PixResponse>(`/v1/aluno/financeiro/${cobrancaId}/pix`).then((r) => r.data),

  getPixStatus: (paymentId: string) =>
    alunoClient.get<PixStatusResponse>(`/v1/aluno/financeiro/pix/${paymentId}/status`).then((r) => r.data),
}
